proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,146 @@
+
+> ```haskell
+> type ALens s t a b = LensLike (Pretext (->) a b) s t a b 
+> ```
+>
+> This type can also be used when you need to store a `Lens` in a container, since it is rank-1.
+
+This proposal makes most of these `newtype` wrappers unnecessary.
+
+Furthermore, if we want to replace an expression such as `view l` with `view $ l`, we need to instantiate the types of `($) :: forall a b. (a -> b) -> a -> b` impredicatively (this case is so common that GHC contains a special case for them).
+
+## Proposed Change Specification
+
+This proposal is structured in two parts. A precise specification can be found in this [paper draft](https://www.microsoft.com/en-us/research/publication/a-quick-look-at-impredicativity/), which would most surely lead to an actual paper submitted to a research conference. Below you can find a more approachable description, which could ultimately lead to a section in the GHC Users Guide.
+
+Any programmer who does not enable `ImpredicativeTypes` is unaware of the contents of this proposal: impredicative instantiation is *not* allowed (rank-n types do not contradict that statement, since there is no instantiation going on during their type checking). In particular, type checking an application `e_0 e_1 ... e_n` in done in the following steps:
+
+1. *Infer* the type of `e_0`, which we shall call `sigma_0`,
+2. Expose the first *n* argument types by instantiating `sigma_0` into a type of the form `sigma_1 -> ... -> sigma_n -> sigma_result`,
+3. *Check* each argument expression *e_i* against the corresponding type `sigma_i`,
+4. If we are in checking mode, run the subtype check (`tcSubType`) of `sigma_result` against the type being pushed.
+
+For example, suppose we are inferring the type of `id True`:
+
+1. We infer the type of `id` by merely looking up in the environment, `forall a. a -> a`,
+2. We expose one argument, and for that we instantiate the type above with a new type variable, leading to `alpha -> alpha`,
+3. Now we check the argument `True` against `alpha`:
+   a. We infer the type of `True`, which is `Bool`,
+   b. Steps (2) and (3) are not needed here, since there are no arguments,
+   c. We perform the subtype check `alpha <= Bool`, leading to the unification `alpha := Bool`;
+4. The result of the inference is `alpha`; after zonking the result is `Bool`.
+
+When `ImpredicativeTypes` is on we introduce an additional step between (2) and (3), which we call *quick look*. Quick look traverses the arguments of the application, and tries to infer as much impredicative instantiations as possible. The results are applied before step (3), which means that arguments are type checked with the correct impredicative instantiations. The "quick look" pass only traverses *simple expressions*, which we define to be (possibly nested) applications of variables in the environment.
```

Comment:
What's missing here is a summary of how the quick-look pass infers impredicative instantiations. It's find to punt to the paper for details. But, as a programmer, I shouldn't need to refer to the paper to build an intuitive understanding of the typing rules.

I think it's reasonable to apply the rule of thumb that the proposal should contain enough information for a programmer to use as its main documentation.

Reply:
I had tried to describe the workings intuitively as "two passes": one over simple expressions where impredicativity can be inferred, and a second one where everything is checked, but impredicativity can no longer be inferred.

@aspiwack @simonpj maybe you could help me in rewording this part?